server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -761,28 +763,44 @@ public Pool find(String poolId) {
         Date entitleDate = data.getOnDate();
         Owner owner = consumer.getOwner();
 
-        List<Entitlement> entitlements = new LinkedList<Entitlement>();
+        int retries = MAX_ENTITLE_RETRIES;
 
-        List<PoolQuantity> bestPools = getBestPools(consumer, productIds,
-            entitleDate, owner, null, fromPools);
+        while (true) {
+            try {
+                List<Entitlement> entitlements = new LinkedList<Entitlement>();
 
-        if (bestPools == null) {
-            List<String> fullList = new ArrayList<String>();
-            fullList.addAll(Arrays.asList(productIds));
-            for (ConsumerInstalledProduct cip : consumer.getInstalledProducts()) {
-                fullList.add(cip.getId());
+                List<PoolQuantity> bestPools = getBestPools(consumer, productIds,
+                    entitleDate, owner, null, fromPools);
+
+                if (bestPools == null) {
+                    List<String> fullList = new ArrayList<String>();
+                    fullList.addAll(Arrays.asList(productIds));
+                    for (ConsumerInstalledProduct cip : consumer.getInstalledProducts()) {
+                        fullList.add(cip.getId());
+                    }
+                    log.info("No entitlements available for products: {}", fullList);
+                    return null;
+                }
+
+                // now make the entitlements
+                for (PoolQuantity entry : bestPools) {
+                    entitlements.add(addOrUpdateEntitlement(consumer, entry.getPool(),
+                        null, entry.getQuantity(), false, CallerType.BIND));
+                }
+
+                return entitlements;
             }
-            log.info("No entitlements available for products: {}", fullList);
-            return null;
-        }
+            catch (EntitlementRefusedException e) {
+                String key = e.getResult().getErrors().get(0).getResourceKey();
 
-        // now make the entitlements
-        for (PoolQuantity entry : bestPools) {
-            entitlements.add(addOrUpdateEntitlement(consumer, entry.getPool(),
-                null, entry.getQuantity(), false, CallerType.BIND));
-        }
+                if (key.equals("rulefailed.no.entitlements.available") && retries-- > 0) {
+                    log.info("No entitlements available while attempting to entitle; retrying");
```

Comment:
Could we change this to something more specific:

"Entitlements exhausted between select best pools and bind operations, retrying"

If we hit this, I believe that's what it has to mean as we check quantity during select best pools.